[BUG][TESTS] Add unit tests for conv asm 1x1u bwdwrw3x3 3x3u 1x1uv2

scerzh opened the pull request:
Adding unit tests for the solvers affected by #3706

scerzh committed:
Fix ConvAsm1x1UV2 applicability issue

scerzh committed:
Fix ConvAsmBwdWrW3x3 applicability issue

scerzh committed:
Fix ConvAsm1x1U applicability issue

scerzh committed:
Add unit test for ConvAsm1x1UV2

scerzh committed:
Add unit test for ConvAsm3x3U

scerzh committed:
Add unit test for ConvAsmBwdWrW3x3

scerzh committed:
Add unit test for ConvAsm1x1U

scerzh committed:
Remove stale conv asm tests

scerzh committed:
Fix a misprint

averinevg commented on src/solver/conv/conv_asm_1x1u.cpp:
```diff
@@ -550,7 +550,7 @@ bool ConvAsm1x1U::IsApplicable(const ExecutionContext& ctx, const ProblemDescrip
 
     const std::string name = ctx.GetStream().GetDeviceName();
-    if(name.find("gfx9") == std::string::npos)
+    if(!(StartsWith(name, "gfx90")))
```
Should `gfx942` be added here or are the necessary instructions missing?

averinevg commented on src/solver/conv/conv_asm_1x1u_stride2.cpp:
```diff
@@ -511,7 +511,7 @@ bool ConvAsm1x1UV2::IsApplicable(const ExecutionContext& ctx,
 
     const std::string name = ctx.GetStream().GetDeviceName();
-    if(name.find("gfx8") == std::string::npos && name.find("gfx9") == std::string::npos)
+    if(!(StartsWith(name, "gfx8") || StartsWith(name, "gfx90")))
```
It might be a good idea to clean up and remove all the code for `gfx8`. Not in this PR of course.

averinevg commented on test/gtest/unit_conv_solver_ConvAsm1x1U.cpp:
```diff
@@ -0,0 +1,185 @@
+    return params;
+}
+
+const auto& GetTestParamsFp16AltEnabled()
+{
+    static const auto params = [] {
+        auto p =
+            miopen::unit_tests::UnitTestConvSolverParams(Gpu::gfx900 | Gpu::gfx906 | Gpu::gfx908);
+        p.Tunable(5);
+        p.SetConvAttrFp16Alt(1);
+        p.CheckXnackDisabled();
+        p.EnableDeprecatedSolvers();
+        return p;
+    }();
+    return params;
+}
```
Note: `AttrFp16Alt` works only for `gfx90a`.

scerzh replied:
Does it mean it should be ignored by any solver on any device other than `gfx90a`?

scerzh replied:
I've removed unnecessary tests.

averinevg commented on test/gtest/unit_conv_solver_ConvAsm1x1U.cpp:
```diff
@@ -0,0 +1,185 @@
+                                          testing::Values(miopenConvolutionAlgoDirect),
+                                          testing::ValuesIn(GetConvTestCases(miopenFloat))));
+
+INSTANTIATE_TEST_SUITE_P(Smoke,
+                         CPU_UnitTestConvSolverAsm1x1UDevApplicabilityFwd_FP16,
+                         testing::Combine(testing::Values(GetTestParamsFp16AltDisabled(),
+                                                          GetTestParamsFp16AltEnabled()),
+                                          testing::Values(GetConvTestCases(miopenHalf)[0])));
+
+INSTANTIATE_TEST_SUITE_P(Smoke,
+                         CPU_UnitTestConvSolverAsm1x1UDevApplicabilityBwd_FP16,
+                         testing::Combine(testing::Values(GetTestParamsFp16AltDisabled(),
+                                                          GetTestParamsFp16AltEnabled()),
+                                          testing::Values(GetConvTestCases(miopenHalf)[0])));
+
+INSTANTIATE_TEST_SUITE_P(Smoke,
+                         CPU_UnitTestConvSolverAsm1x1UDevApplicabilityFwd_FP32,
+                         testing::Combine(testing::Values(GetTestParamsFp32()),
+                                          testing::Values(GetConvTestCases(miopenFloat)[0])));
+
+INSTANTIATE_TEST_SUITE_P(Smoke,
+                         CPU_UnitTestConvSolverAsm1x1UDevApplicabilityBwd_FP32,
+                         testing::Combine(testing::Values(GetTestParamsFp32()),
+                                          testing::Values(GetConvTestCases(miopenFloat)[0])));
```
There is no need to run this test for all data types and directions. Although it is a short test and has almost no impact on the overall test time, it doesn't improve test coverage either.

scerzh replied:
Backward and forward path have different default values for `Fp16Alt` flag and the result of `IsApplicable()` depends on it. But if the flag should be ignored on any device except `gfx90a`, 2/3 of these tests are useless and it's only needed to confirm the solver isn't applicable for `gfx90a` with the flag set.

https://github.com/ROCm/MIOpen/blob/5b9d08fb31e1c1ccd1a784b3c63499bee7c589f7/src/include/miopen/convolution.hpp#L83-L85

scerzh replied:
I've removed unnecessary tests.

scerzh committed:
Remove unnecessary test cases

scerzh committed:
Merge develop into sg/add_unit_tests_conv_asm_1x1u_bwdwrw3x3_3x3u_1x1uv

scerzh committed:
Add an exception to gtest_formating_checks

gtest_formating_checks doesn't understand preprocessor language and
incorrectly detects test-suit double-instantiation

shurale-nkn committed:
Merge branch 'develop' into sg/add_unit_tests_conv_asm_1x1u_bwdwrw3x3_3x3u_1x1uv

scerzh committed:
Merge branch 'develop' into sg/add_unit_tests_conv_asm_1x1u_bwdwrw3x3_3x3u_1x1uv

BrianHarrisonAMD committed:
Merge branch 'develop' into sg/add_unit_tests_conv_asm_1x1u_bwdwrw3x3_3x3u_1x1uv

BrianHarrisonAMD reviewed (approved):
LGTM assuming tests pass.

scerzh committed:
Merge branch 'develop' into sg/add_unit_tests_conv_asm_1x1u_bwdwrw3x3_3x3u_1x1uv
